### DISCUSSION TITLE
My experience with Next.js 14 app router

### DISCUSSION THREAD

**Discussion by ersinakinci** [mikro-orm/mikro-orm#5467] [General]
Just sharing my experiences in the hope that they would be helpful to someone else.

I nearly tore my hair out trying to get MikroORM to work for a new web app using Next.js 14's app router. I forget all the approaches that I tried, but it didn't work. All of the repos and suggestions I looked at were outdated. I had to hack very deep into Next.js and MikroORM to diagnose what was broken.

There are the usual suspects (externals broken, etc.), many of which are addressed in the [deployment docs](https://mikro-orm.io/docs/deployment). If you go down the Next.js rabbit hole, you'll run into them and can use that doc to solve them.

But the core, nearly unsolvable problem is that Next.js uses SWC to compile JS/TS. SWC by default mangles all the code and this interferes with MikroORM's metadata providers. SWC also provides settings to prevent mangling, but 1) these aren't exposed by Next.js, and 2) even if they were, there's no easy way to say "don't mangle the server-side code but do mangle the client JS," which means that you're going to have lower-performing, bloated JS sent to web browsers.

If you're absolutely bent on this path, I managed to get everything working, but I don't recommend it.

First, you need to set up `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    serverComponentsExternalPackages: ["umzug", "jju"],
  },
  swcMinify: false,
};

module.exports = nextConfig;
```

Next will complain loudly about dropping support for `swcMinify` in the next major release. That option makes Next.js minify using the older Terser library instead of SWC. I didn't think this setting would have any effect, but I found in practice that it was necessary for the following step, which is where it gets ugly.

After configuring `next.config.js`, you're gonna need to patch the Next distributed files to override their hardcoded SWC config. How you do this is up to you. If you're using `npm`, you can probably use something like [patch-package](https://github.com/ds300/patch-package). I used `pnpm patch` to modify `./node_modules/next/dist/build/swc/options.js`. The things I changed were:

1. Change `decorators: enableDecorators` to `decorators: true` in the options returned from `getParserOptions`
2. Add `minify: false, jsc: { ..., keepClassNames: true, minify: { mangle: false, compress: false } }` to the options returned from `getBaseSWCOptions`
3. Change `jsc: { transform: { legacyDecorator: enableDecorators, decoratorMetadata: emitDecoratorMetadata } }` to `jsc: { transform: { legacyDecorator: true, decoratorMetadata: true } }` in the options returned from `getBaseSWCOptions`.

It worked, but the experience was so janky that in the end I gave up and used Remix for my web app. Remix uses vanilla Vite for bundling and the dev server so it's much easier to configure to be compatible with MikroORM.

**Comment by kchojhu** [comment]
Thanks for this but is this still the case for NextJs to use experimental features?

**Comment by andres-dejesus** [comment]
Any updates? Is this still the only way to get Mikro to work with Nextjs 14?

**Reply by ersinakinci** [reply]
No updates from me. As I mentioned in the post, I switched to Remix, and eventually I switched from MikroORM to Kysely.

**Reply by andres-dejesus** [reply]
For anyone else interested, I got it work by setting the following on next.config :

 ```
experimental: { serverMinification: false }
```

Issue is being tracked here:  https://github.com/vercel/next.js/issues/59594.

**Reply by jacobtipp** [reply]
> For anyone else interested, I got it work by setting the following on next.config :
> 
> ```
> experimental: { serverMinification: false }
> ```
> 
> Issue is being tracked here: [vercel/next.js#59594](https://github.com/vercel/next.js/issues/59594).

I don't think this ideal as the docker build would be large in production among other problems.

**Comment by gugupy** [comment]
Hi,

I am new to NextJS and TypeScript and want to building an application using NextJS and MikroORM. 

But stucked with creating monorepo and tsconfig conflicts, any helps would be appreciated. 

Thanks
